### PR TITLE
A

### PR DESCRIPTION
**Pull request by demipixel**
is for behAviors!

### CODE REVIEW

**Comment by 109C** [comment]
behAviators?

File: src/lib/plugins/sound.js [hunk 1]
```diff
@@ -40,11 +40,12 @@ module.exports.player=function(player,serv) {
     serv.playSound(sound, player.world, null, opt);
   };
 
-  player._client.on('block_place', ({location}={}) => {
+  player._client.on('placeBlock_cancel', ({location}={}, cancel) => {
```
Comment (rom1504) [inline src/lib/plugins/sound.js]: On _client really ?

Comment (demipixel) [inline src/lib/plugins/sound.js]: IM SRY AFTER 650 LINES U HAVE TO UNDERSTAND PLEASE FORGIVE

**Comment by mhsjlw** [comment]
This is the worst so far. :monkey:

File: src/lib/plugins/digging.js [hunk 2]
```diff
@@ -2,48 +2,46 @@ var Vec3 = require("vec3").Vec3;
...
-      )}
-    )
+        currentlyDugBlock=block;
+        if(currentlyDugBlock.type==0) return;
+        if(status==0 && player.gameMode!=1)
+          player.behavior('dig', { // Start dig survival
+            position: pos,
+            block: block
+          }, ({position}) => {
+            return startDigmehging(position);
```
Comment (rom1504) [inline src/lib/plugins/digging.js]: startDigmehging ???

File: src/lib/plugins/sound.js [hunk 3]
```diff
@@ -40,29 +40,28 @@ module.exports.player=function(player,serv) {
...
-  player._client.on('block_dig', ({location,status} = {}) => {
+  player.on('dig_cancel', async ({position, reference}, cancel) => {
     if (status != 0 || player.gameMode == 1) return;
-    var pos=new Vec3(location.x,location.y,location.z);
-    player.world.getBlockType(pos).then((id) => {
+    return player.world.getBlockType(reference).then((id) => {
       if (id != 25) return;
-      if (!player.world.blockEntityData[pos.toString()]) player.world.blockEntityData[pos.toString()] = {};
-      var data = player.world.blockEntityData[pos.toString()];
+      cancel();
```
Comment (rom1504) [inline src/lib/plugins/sound.js]: not cancel(false); ?

File: doc/api.md [hunk 4]
```diff
@@ -343,6 +601,148 @@ Fires when the player says `message`.
...
+- sound: Sound to play (Default: default sound for that material)
+
+Default: Place block for server and nearby players
+
+Cancelled: Replace block with old block for player
+
+#### "attack"
+
+Emitted when a player attacks an entity
+- attackdEntity: Entity being attacked
```
Comment (rom1504) [inline doc/api.md]: attackdEntity

**Comment by rom1504** [comment]
I made a few comments but basically it's okay, is everything working ?

**Comment by demipixel** [comment]
Note blocks aren't but we need the extended EventEmitter for that.

**Comment by rom1504** [comment]
@demipixel why ?

**Comment by demipixel** [comment]
@rom1504 We need to do a block-get to check if it's a noteblock that they're right-clicking. The current behavior system doesn't handle async.

**Comment by rom1504** [comment]
okay,
so just got to use that https://www.npmjs.com/package/emit-then then

**Comment by rom1504** [comment]
got to put emit-then in package.json ;)

**Pull request merged**